rust/operator-binary/src/history_controller.rs

```diff
@@ -0,0 +1,511 @@
+            .with_recommended_labels(labels(shs, app_version_label, &group_name))
+            .build(),
+        spec: Some(ServiceSpec {
+            ports: Some(vec![ServicePort {
+                name: Some(String::from("http")),
+                port: 18080,
+                ..ServicePort::default()
+            }]),
+            selector: Some(selector),
+            type_: Some("NodePort".to_string()),
+            ..ServiceSpec::default()
+        }),
+        status: None,
+    })
+}
+
+fn build_history_role_serviceaccount(
+    shs: &SparkHistoryServer,
+    app_version_label: &str,
+) -> Result<(ServiceAccount, RoleBinding)> {
+    let sa_name = shs.metadata.name.as_ref().unwrap().to_string();
+    let sa = ServiceAccount {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(shs)
+            .name(&sa_name)
+            .ownerreference_from_resource(shs, None, Some(true))
+            .context(ObjectMissingMetadataForOwnerRefSnafu)?
+            .with_recommended_labels(labels(shs, app_version_label, HISTORY_CONTROLLER_NAME))
+            .build(),
+        ..ServiceAccount::default()
+    };
+    let binding_name = &sa_name;
+    let binding = RoleBinding {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(shs)
+            .name(binding_name)
+            .ownerreference_from_resource(shs, None, Some(true))
+            .context(ObjectMissingMetadataForOwnerRefSnafu)?
+            .with_recommended_labels(labels(shs, app_version_label, HISTORY_CONTROLLER_NAME))
+            .build(),
+        role_ref: RoleRef {
+            api_group: <ClusterRole as stackable_operator::k8s_openapi::Resource>::GROUP // need to fully qualify because of "Resource" name clash
+                .to_string(),
+            kind: <ClusterRole as stackable_operator::k8s_openapi::Resource>::KIND.to_string(),
+            name: SPARK_CLUSTER_ROLE.to_string(),
+        },
+        subjects: Some(vec![Subject {
+            api_group: Some(
+                <ServiceAccount as stackable_operator::k8s_openapi::Resource>::GROUP.to_string(),
+            ),
+            kind: <ServiceAccount as stackable_operator::k8s_openapi::Resource>::KIND.to_string(),
+            name: sa_name,
+            namespace: sa.metadata.namespace.clone(),
+        }]),
+    };
+    Ok((sa, binding))
+}
+
+fn spark_config(
+    shs: &SparkHistoryServer,
+    s3_log_dir: &S3LogDir,
+    rolegroupref: &RoleGroupRef<SparkHistoryServer>,
+) -> Result<String, Error> {
+    let empty = BTreeMap::new();
+
+    let mut log_dir_settings = s3_log_dir.history_server_spark_config();
+
+    // add cleaner spark settings if requested
+    log_dir_settings.extend(cleaner_config(shs, rolegroupref)?);
+
+    // add user provided configuration. These can overwrite everything.
+    let user_settings = shs.spec.spark_conf.as_ref().unwrap_or(&empty);
+    log_dir_settings.extend(user_settings.clone().into_iter());
```

Comment:
`empty` can be removed
```suggestion
    log_dir_settings.extend(shs.spec.spark_conf.clone().unwrap_or_default());
```